File: c/src/minesweeper.c
#include <ncurses.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include "libmine/include/libmine.h"
#define LOG_USE_COLOR
#define ATD_LOGGER_IMPLEMENTATION
#include "logger.h"

static size_t cur_x = 0;
static size_t cur_y = 0;

bool update(MinesweeperGame** g){
	MinesweeperGame* game = *g;
	switch (getch()) {
		case KEY_UP: {
			cur_y--;
		} break;
		case KEY_DOWN: {
			cur_y++;
		} break;
		case KEY_LEFT: {
			cur_x--;
		} break;
		case KEY_RIGHT: {
			cur_x++;
		} break;
		case 'F':
		case 'f': {
			toggleFlag(game, cur_x, cur_y);
		} break;
		case 'R':
		case 'r': {
			*g = refreshGame(game);
		} break;
		case '\n':
		case KEY_ENTER: {
			if (game->status.state == WAITING_FIRST_MOVE || game->status.state == PLAYING) {
				reveal(game, cur_x, cur_y);
			}
		} break;
		case KEY_EXIT:
		case 'q': {
			return false;
		} break;
		case KEY_MOUSE: {
			MEVENT event;
			if (getmouse(&event) != OK) {
				log_debug("Got bad mouse event.");
				break;
			}
			if(event.bstate & BUTTON1_CLICKED) {
				log_debug("Click.");
			}
			else if(event.bstate & BUTTON3_CLICKED) {
				log_debug("Clock.");
			}
			log_debug("Mouse at row=%d, column=%d bstate=0x%08lx", event.y, event.x, event.bstate);
		} break;
	}
	return true;
}

void render(MinesweeperGame* game){
	attron(COLOR_PAIR(1));
	printw("Ahmet Tarik Duyar \n");
	for (size_t y = 0; y < game->board->height; y++) {
		for (size_t x = 0; x < game->board->width; x++) {
			Cell* c = getCell(game->board, x, y);
			int col = COLOR_PAIR(-1);
			switch (c->status) {
				case OPENED: {
					if (c->value == 'B')
						col = COLOR_PAIR(4) | A_BOLD;
					else if(c->value-'0' < 4)
						col = COLOR_PAIR(c->value-'0' + 5);
					else
						col = COLOR_PAIR(9);
				} break;
				case CLOSE: {
					col = COLOR_PAIR(0);
				} break;
				case FLAGGED: {
					col = COLOR_PAIR(3);
				} break;
			}
			if (x == cur_x && y == cur_y) {
				col = COLOR_PAIR(2) | A_BOLD;
			}
			attron(col);
			switch (c->status) {
				case OPENED: {
					printw(" %c ", c->value);
				} break;
				case CLOSE: {
					printw(" # ");
				} break;
				case FLAGGED: {
					printw(" F ");
				} break;
			}
			attroff(col);
		}
		printw("\n");
	}
	printw("---------STATUS-----------\n");
	printw("Flagged: %d\n", game->status.flagedCellCount);
	printw("Revealed: %d\n", game->status.revealedCellCount);
	printw("Time: %ld\n", time(NULL) - game->status.startTime);
	printw("State: %s\n", game->status.state == YOU_WIN ? "YOU WIN" : game->status.state == YOU_GAMEOVER ? "YOU GAMEOVER" : "PLAYING");
	printw("---------SETTINGS---------\n");
	printw("Width: %zu, Height: %zu, Mines: %u\n", game->settings.width, game->settings.height, game->settings.mines);
	printw("SmartReveal: %s", game->settings.smartReval?"Active":"Disabled");
}

#define COLOR_GRAY_0 8
#define COLOR_GRAY_1 9
#define COLOR_GRAY_2 10
#define COLOR_GRAY_3 11
#define COLOR_GRAY_4 12

static int setColor() {
	if (!has_colors()) {
        endwin();
        log_warn("Your terminal does not support colors\n");
        return 1;
    }
    if (!can_change_color()) {
        endwin();
        log_warn("Your terminal does not support color modifications\n");
        return 1;
    }
    start_color();
	use_default_colors();
    init_color(COLOR_GRAY_0, 50, 50, 50);
    init_color(COLOR_GRAY_1, 200, 200, 200);
    init_color(COLOR_GRAY_2, 200, 200, 500);
    init_color(COLOR_GRAY_3, 500, 200, 300);
    init_color(COLOR_GRAY_4, 700, 200, 200);
    init_pair(1, -1, -1);
    init_pair(2, COLOR_YELLOW, COLOR_BLACK); // cursor
    init_pair(3, COLOR_GREEN, -1); // flag
    init_pair(4, COLOR_RED, -1); // BOMB
    init_pair(5, COLOR_GRAY_0, -1);
    init_pair(6, COLOR_GRAY_1, -1);
    init_pair(7, COLOR_GRAY_2, -1);
    init_pair(8, COLOR_GRAY_3, -1);
    init_pair(9, COLOR_GRAY_4, -1);

	return 0;
}

int main(void) {
	srand(time(NULL));

	log_add_fp(fopen("debug.log", "a"), LOG_DEBUG);
	log_set_quiet(true);
	log_debug("Mineswepeer started.");

	// GameSettings settings = { .difficulty=CUSTOM, .width=10, .height=10, .mines=100 };
	// GameSettings settings = { .difficulty=CUSTOM, .width=1, .height=1, .mines=1, .smartReval=true };
	GameSettings settings = { .difficulty=CUSTOM, .width=10, .height=10, .mines=10, .smartReval=true };
	// GameSettings settings = { .difficulty=CUSTOM, .width=5, .height=5, .mines=4, .smartReval=true };
	// GameSettings settings = { .difficulty= EASY };
	// GameSettings settings = { .difficulty= MEDIUM };
	// GameSettings settings = { .difficulty= HARD };
	MinesweeperGame* mineGame = newGame(settings);

	initscr();
	setColor();

	cbreak();
	noecho();
	keypad(stdscr, TRUE);
	mousemask(ALL_MOUSE_EVENTS | REPORT_MOUSE_POSITION, NULL);
	curs_set(0); 
	bool isRuning = true;
	while (isRuning) {
		clear();
		render(mineGame);
		refresh();
		isRuning = update(&mineGame);
	}
	freeGame(mineGame);
	endwin();
	return 0;
}

File: python/minesweeper.py
from blessed import Terminal
import random

class GameSettings:
    def __init__(self,difficulty,width=None,height=None,mine=None):
        match difficulty:
            case "easy":
                self.width=9
                self.height=9
                self.mine=10
            case "medium":
                self.width=16
                self.height=16
                self.mine=40
            case "hard":
                self.width=30
                self.height=16
                self.mine=99
            case "custom":
                self.height=height
                self.width=width
                self.mine=mine

class Cell:
    def __init__(self,isMine,number=None):
        self.isMine=isMine
        self.number=number
        self.isFlagged=False
        self.isOpened=False
    def __str__(self):
        if self.isFlagged:
            return term.bright_green+term.bold+ ' F '+term.normal
        elif self.isOpened==False:
            return ' # '
        else:
            if self.isMine:
                return term.red+term.bold+" B " + term.normal
            else:
                match self.number:
                    case 0:
                        return "   "
                        #return term.black+" "+ str(self.number) + " "+term.normal
                    case 1:
                        return term.black+" "+ str(self.number) + " "+term.normal
                    case 2:
                        return term.royalblue+" "+ str(self.number) + " "+term.normal
                    case 3:
                        return term.maroon+" "+ str(self.number) + " "+term.normal
                    case _:
                        return term.red+" "+ str(self.number) + " "+term.normal
    
    def setMine(self):
        self.isMine=True
        self.number=None

    def setNumber(self,num):
        self.isMine=False
        self.number=num

class Board:
    def __init__(self,settings):
        self.cursor = [0, 0]
        self.settings=settings
        self.gameStatus="playing"
        self.openedCellsCount=0
        self.isFirstMove=True
        self.cells = []
        # Creating cells
        for y in range(0,self.settings.height):
            self.cells.append([])
            for x in range(0,self.settings.width):
                self.cells[y].append(Cell(False, 0))
                #self.cells[y][x].isOpened=True
        #Creating mines
        for _ in range (0,self.settings.mine):
            self.FindRandomNonMineCell().setMine()

        #Creating numbers
        self.CalcNumbers()
    
    def FindRandomNonMineCell(self):
        while True:
                randX = random.randint(0,self.settings.width-1)
                randY = random.randint(0,self.settings.height-1)
                if self.cells[randY][randX].isMine:
                    continue
                else:
                    return self.cells[randY][randX]

    def CalcNumbers(self):
        for y in range(0,self.settings.height):
            for x in range(0,self.settings.width):
                if self.cells[y][x].isMine:
                    continue
                mineCount=0
                for ry in range(max(0,y-1),min(y+2,self.settings.height)):
                    for rx in range(max(0,x-1),min(x+2,self.settings.width)):
                        if self.cells[ry][rx].isMine:
                            mineCount+=1
                self.cells[y][x].setNumber(mineCount)

    def render(self):
        for y in range (0,self.settings.height):
            for x in range(0,self.settings.width):
                if x == self.cursor[0] and y == self.cursor[1]:
                    print(term.on_snow(str(self.cells[y][x])), end="")
                else:
                    print(self.cells[y][x], end="")
            print()
    
    def moveCursorUp(self):
        if self.cursor[1] > 0:
            self.cursor[1] -= 1
    def moveCursorDown(self):
        if self.cursor[1] < self.settings.height-1:
            self.cursor[1] += 1
    def moveCursorLeft(self):
        if self.cursor[0] > 0:
            self.cursor[0] -= 1
    def moveCursorRight(self):
        if self.cursor[0] < self.settings.width-1:
            self.cursor[0] += 1
    
    def toggleFlag(self):
        if self.gameStatus != "playing":
            return
        
        cell = self.cells[self.cursor[1]][self.cursor[0]]
        if not cell.isOpened:
            cell.isFlagged= not cell.isFlagged
    def reveal(self, x=None,y=None):
        if self.gameStatus != "playing":
            return
        
        if x == None or y == None:
            x = self.cursor[0]
            y = self.cursor[1]
        
        cell = self.cells[y][x]
            
        if cell.isFlagged:
            return
        if cell.isOpened:
            if cell.number != 0:
                flaggedCount = 0
                for ry in range(max(0,y-1),min(y+2,self.settings.height)):
                    for rx in range(max(0,x-1),min(x+2,self.settings.width)):
                        rCell = self.cells[ry][rx]
                        if rCell.isFlagged:
                            flaggedCount += 1
                if flaggedCount == cell.number:
                    for ry in range(max(0,y-1),min(y+2,self.settings.height)):
                        for rx in range(max(0,x-1),min(x+2,self.settings.width)):
                            rCell = self.cells[ry][rx]
                            if not rCell.isFlagged and not rCell.isOpened:
                                self.reveal(rx, ry)
            return
        cell.isOpened = True
        
        if cell.isMine:
            if self.isFirstMove:
                randCell = self.FindRandomNonMineCell()
                randCell.setMine()
                cell.isMine = False
                self.CalcNumbers()
            else:
                self.gameStatus="gameOver"
                for y in range(0,self.settings.height):
                    for x in range(0,self.settings.width):
                        cell = self.cells[y][x]
                        if cell.isMine:
                            cell.isOpened = True
                            cell.isFlagged = False
                return
        if cell.number == 0:
            for ry in range(max(0,y-1),min(y+2,self.settings.height)):
                for rx in range(max(0,x-1),min(x+2,self.settings.width)):
                    cell = self.cells[ry][rx]
                    self.reveal(rx, ry)
        
        self.isFirstMove = False
        self.openedCellsCount += 1
        if board.settings.height*board.settings.width - board.settings.mine == self.openedCellsCount:
            self.gameStatus = "gameWin"
        

term = Terminal()
isMenu = True
board=Board(GameSettings("hard"))
        
while True:
    # render
    if isMenu:
        print(term.home + term.clear)
        print(term.black_on_darkkhaki(term.center('Minesweeper')))
        print("""  __  __ _             _____                                  
 |  \/  (_)           / ____|                                 
 | \  / |_ _ __   ___| (_____      _____  ___ _ __   ___ _ __ 
 | |\/| | | '_ \ / _ \\___ \ \ /\ / / _ \/ _ \ '_ \ / _ \ '__|
 | |  | | | | | |  __/____) \ V  V /  __/  __/ |_) |  __/ |   
 |_|  |_|_|_| |_|\___|_____/ \_/\_/ \___|\___| .__/ \___|_|   
                                             | |              
                                             |_|              """)
        
        print()
        print(term.black_on_darkkhaki(term.center("Choose difficulty:")))
        print("\t1: easy")
        print("\t2: medium")
        print("\t3: hard")
        print()
        print("\tq: quit")
        with term.cbreak(), term.hidden_cursor():
            inp = term.inkey()
        match inp:
            case "1":
                board=Board(GameSettings("easy"))
                isMenu = False
            case "2":
                board=Board(GameSettings("medium"))
                isMenu = False
            case "3":
                board=Board(GameSettings("hard"))
                isMenu = False
            case "q":
                exit(0)
        continue
    
    print(term.home + term.clear)
    print(term.black_on_darkkhaki(term.center('Minesweeper')))

    match board.gameStatus:
        case "playing":
            print(term.black_on_yellow(term.center(f"{board.openedCellsCount}/{board.settings.height*board.settings.width-board.settings.mine}")))
        case "gameWin":
            print(term.black_on_green(term.center('You Win! Press "r" to replay')))
        case "gameOver":
            print(term.black_on_red(term.center('Game Over Press "r" to replay')))
    board.render()
        
    print(term.black_on_darkkhaki(term.center('Status')))
    print(end="", flush=True)
    # update
    with term.cbreak(), term.hidden_cursor():
        inp = term.inkey()
    if inp.is_sequence:
        match inp.name:
            case "KEY_UP":
                board.moveCursorUp()
            case "KEY_DOWN":
                board.moveCursorDown()
            case "KEY_LEFT":
                board.moveCursorLeft()
            case "KEY_RIGHT":
                board.moveCursorRight()
            case "KEY_ENTER":
                board.reveal()
    else:
        match inp:
            case "F" | "f":
                board.toggleFlag()
            case "r":
                board=Board(board.settings)
            case " ":
                board.reveal()
            case "q":
                isMenu = True
